python/ray/autoscaler/_private/vsphere/node_provider.py

```
@@ -0,0 +1,726 @@
+        yn_id = DynamicID(type=Constants.TYPE_OF_RESOURCE, id=vm_id)
+
+        # List all the tags present on the VM.
+        for tag_id in self.vsphere_sdk_client.tagging.TagAssociation.list_attached_tags(
+            yn_id
+        ):
+            vsphere_vm_tag = self.vsphere_sdk_client.tagging.Tag.get(tag_id=tag_id).name
+            tag_key_value = vsphere_tag_to_kv_pair(vsphere_vm_tag)
+            tag_key = tag_key_value[0] if tag_key_value else None
+            if tag_key == tag_key_to_remove:
+                # Remove the tag matching the key passed.
+                cli_logger.info("Removing tag {} from the VM {}".format(tag_key, vm_id))
+                self.vsphere_sdk_client.tagging.TagAssociation.detach(tag_id, yn_id)
+                break
+
+    def get_frozen_vm_obj(self):
+        frozen_vm_target_name = self.frozen_vm_name
+        vms = self.vsphere_sdk_client.vcenter.VM.list(
+            VM.FilterSpec(names={frozen_vm_target_name})
+        )
+
+        cli_logger.info("VM to clone from {}".format(vms))
+
+        return vms[0] if len(vms) > 0 else None
+
+    def get_pyvmomi_obj(self, vimtype, name):
+        obj = None
+
+        # TODO: Find a better way to solve pyvmomi timeout issues
+        self.vsphere_pyvmomi_sdk_client = VmwSdkClient(
+            self.vsphere_credentials["server"],
+            self.vsphere_credentials["user"],
+            self.vsphere_credentials["password"],
+            VmwSdkClient.SessionType.UNVERIFIED,
+            VmwSdkClient.ClientType.PYVMOMI_SDK,
+        ).get_client()
+
+        container = self.vsphere_pyvmomi_sdk_client.viewManager.CreateContainerView(
+            self.vsphere_pyvmomi_sdk_client.rootFolder, vimtype, True
+        )
+
+        for c in container.view:
+            if name:
+                if c.name == name:
+                    obj = c
+                    break
+            else:
+                obj = c
+                break
+
+        return obj
```

In case `container.view` is empty, do the callers of `get_pyvmomi_obj` handle receiving the return value `None` appropriately, or is it better to raise an exception? (Or do we not expect this to happen in practice)

---

I have added the logic to raise an exception. We expect the obj to be fetchable with the object name.